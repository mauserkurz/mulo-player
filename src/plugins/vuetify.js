import Vue from 'vue';
import Vuetify from 'vuetify';
import { THEME as theme } from '@/constants';

Vue.use(Vuetify);

export default new Vuetify({ theme });
