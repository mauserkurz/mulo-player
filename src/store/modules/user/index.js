import { path, pathOr } from 'ramda';
import utils from '@/utils';
import api from '@/api/';
import router from '@/router';
import { AUTH_FORM_TYPE_MAP, STATUS_MAP, API_ERROR_MAP } from '@/const';

const getErrorKey = response => pathOr([], ['data', 'errors', 0, 'message'], response);

export default {
  namespaced: true,

  state: {
    userID: '',
    login: '',
    error: '',
    isAuthLoading: false,
    isUserDataLoading: false,
    isLogoutProcessing: false,
  },

  getters: {
    isAuthorized(state) {
      return !!state.userID;
    },
  },

  mutations: {
    SET_USER_ID(state, id) {
      state.userID = id;
    },

    SET_LOGIN(state, login) {
      state.login = login;
    },

    SET_AUTH_ERROR(state, error) {
      state.error = error;
    },

    SET_IS_AUTH_LOADING(state, mode) {
      state.isAuthLoading = mode;
    },

    SET_IS_USER_DATE_LOADING(state, mode) {
      state.isUserDataLoading = mode;
    },

    SET_IS_LOGOUT_PROCESSING(state, mode) {
      state.isLogoutProcessing = mode;
    },
  },

  actions: {
    setUser({ commit }, data) {
      commit('SET_USER_ID', data.user_id);
      commit('SET_LOGIN', data.login);
      router.push({ path: 'player' });
    },

    async auth({ commit, dispatch }, { login, password, type }) {
      const methodMap = {
        [AUTH_FORM_TYPE_MAP.SIGN_IN]: 'signIn',
        [AUTH_FORM_TYPE_MAP.SIGN_UP]: 'signUp',
      };
      let response;

      commit('SET_IS_AUTH_LOADING', true);
      try {
        const request = api[methodMap[type]];

        response = await request({ login, password });
      } catch (error) {
        const errorKey = getErrorKey(error.response);
        const message = API_ERROR_MAP.AUTH[errorKey];

        // For 401 status show message from map
        // otherwise show message generated from error
        if (path(['response', 'status'], error) === STATUS_MAP.UNAUTHORIZED && message) {
          commit('SET_AUTH_ERROR', message);
        } else {
          commit('SET_AUTH_ERROR', utils.createMessage(error));
        }
        return;
      } finally {
        commit('SET_IS_AUTH_LOADING', false);
      }
      // After successful request clear error, write user id and login
      // then show player itself
      commit('SET_AUTH_ERROR', '');
      dispatch('setUser', response.data);
    },

    async getUser({ commit, dispatch }) {
      let response;

      commit('SET_IS_USER_DATE_LOADING', true);
      try {
        response = await api.getUser();
      } catch (error) {
        if (path(['response', 'status'], error) !== STATUS_MAP.UNAUTHORIZED) {
          window.console.error(error);
        }
        return;
      } finally {
        commit('SET_IS_USER_DATE_LOADING', false);
      }
      dispatch('setUser', response.data);
    },

    async logout({ state, commit }) {
      if (state.isLogoutProcessing) {
        return;
      }
      commit('SET_IS_LOGOUT_PROCESSING', true);

      try {
        await api.logout();
      } catch (error) {
        if (path(['response', 'status'], error) !== STATUS_MAP.UNAUTHORIZED) {
          window.console.error(error);
        }
        commit('SET_IS_LOGOUT_PROCESSING', false);
        return;
      }

      commit('SET_USER_ID', '');
      commit('SET_LOGIN', '');
      await router.push({ path: '/' });
      commit('SET_IS_LOGOUT_PROCESSING', false);
    },

    clearError({ commit }) {
      commit('SET_AUTH_ERROR', '');
    },
  },
};
