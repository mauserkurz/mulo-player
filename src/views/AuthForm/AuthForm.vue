<template>
  <v-card class="auth-form elevation-12">
    <v-form
      ref="form"
      @submit.prevent="submit">
      <v-toolbar
        color="primary"
        dark
        flat>
        <v-toolbar-title>{{ title }}</v-toolbar-title>
        <v-spacer/>
        <v-btn
          text
          small
          class="auth-form__sign-in-tab"
          @click="toggleType">
          Sign In
        </v-btn>
        <v-btn
          text
          small
          class="auth-form__sign-up-tab"
          @click="toggleType">
          Sign Up
        </v-btn>
      </v-toolbar>
      <v-card-text>
        <v-text-field
          :error="hasError"
          :rules="loginRules"
          label="Login"
          name="login"
          prepend-icon="person"
          type="email"
          v-model="login"/>

        <v-text-field
          :error="hasError"
          :rules="passwordRules"
          label="Password"
          name="password"
          prepend-icon="lock"
          type="password"
          v-model="password"/>
      </v-card-text>

      <v-expand-transition>
        <v-card-text
          v-show="hasError">
          <v-alert
            type="error"
            dense>
            {{ error }}
          </v-alert>
        </v-card-text>
      </v-expand-transition>

      <v-card-actions>
        <v-spacer/>
        <v-btn
          :loading="loading"
          type="submit"
          color="primary"
          class="auth-form__submit-button">
          {{ submitButtonText }}
        </v-btn>
      </v-card-actions>
    </v-form>
  </v-card>
</template>

<script>
import { AUTH_FORM_TYPE_MAP, EMAIL_REGEXP, PASSWORD_REGEXP } from '@/constants';

export default {
  name: 'AuthForm',

  props: {
    error: {
      type: String,
      default: '',
    },

    loading: {
      type: Boolean,
      default: false,
    },
  },

  data() {
    return {
      login: '',
      password: '',
      type: AUTH_FORM_TYPE_MAP.SIGN_IN,
      loginRules: [
        value => !!value || 'Please enter e-mail',
        value => EMAIL_REGEXP.test(value) || 'Incorrect e-mail, try again',
      ],
      passwordRules: [
        value => !!value || 'Please enter password',
        value => value.length === 6 || 'Password should contain 6 characters',
        value => PASSWORD_REGEXP.test(value)
          || 'Password should have only latin symbols and numbers',
      ],
    };
  },

  computed: {
    hasError() {
      return !!this.error;
    },

    isSignUp() {
      return this.type === AUTH_FORM_TYPE_MAP.SIGN_UP;
    },

    title() {
      return this.isSignUp ? 'Registration form' : 'Login form';
    },

    submitButtonText() {
      return this.isSignUp ? 'Submit' : 'Login';
    },
  },

  methods: {
    toggleType() {
      const type = this.isSignUp ? AUTH_FORM_TYPE_MAP.SIGN_IN : AUTH_FORM_TYPE_MAP.SIGN_UP;

      this.type = type;
      this.$emit('toggle-type', type);
    },

    submit() {
      const { login, password, type } = this;

      if (this.$refs.form.validate()) {
        this.$emit('submit', { login, password, type });
      }
    },
  },
};
</script>

<style scoped lang="less">
.auth-form {}
</style>
