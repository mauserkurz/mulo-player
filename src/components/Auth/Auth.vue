<template>
  <v-main class="auth">
    <v-container
      class="fill-height"
      fluid>
      <v-row
        align="center"
        justify="center">
        <v-col
          cols="12"
          sm="8"
          md="4">
          <v-row
            v-if="isUserDataLoading"
            align="center"
            justify="center">
            <v-progress-circular
              :size="60"
              indeterminate
              :color="$vuetify.theme.themes.light.primary"/>
          </v-row>

          <AuthForm
            v-else
            :loading="isAuthLoading"
            :error="error"
            @submit="auth"
            @toggle-type="clearError"/>
        </v-col>
      </v-row>
    </v-container>
  </v-main>
</template>

<script>
import { mapState, mapActions } from 'vuex';
import AuthForm from '@/views/AuthForm/AuthForm.vue';

export default {
  name: 'Auth',

  components: { AuthForm },

  computed: {
    ...mapState('user', ['error', 'isAuthLoading', 'isUserDataLoading']),
  },

  methods: {
    ...mapActions('user', ['auth', 'getUser', 'clearError']),
  },

  async mounted() {
    await this.getUser();
  },
};
</script>

<style scoped lang="less">
.auth {}
</style>
