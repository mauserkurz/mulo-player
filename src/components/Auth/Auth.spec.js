import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import clone from 'ramda/src/clone';
import Vuex from 'vuex';
import { toMatchDiffSnapshot } from 'snapshot-diff';
import vuetify from '@/plugins/vuetify';
import user from '@/store/modules/user/';
import AuthForm from '@/views/AuthForm/AuthForm.vue';
import Auth from './Auth.vue';

const localVue = createLocalVue();

localVue.use(Vuex);
expect.extend({ toMatchDiffSnapshot });

describe('Component Auth', () => {
  const createWrapper = ({ isWithoutStubs = false, modules = {}, ...options } = {}) => {
    const renderer = isWithoutStubs ? mount : shallowMount;
    const store = new Vuex.Store({
      namespaced: true,

      modules: Object.assign({ user }, modules),
    });
    const wrapper = renderer(Auth, {
      localVue,
      vuetify,
      store,
      ...options,
    });

    return { wrapper, store };
  };
  let originGetUser;

  beforeAll(() => {
    originGetUser = user.getUser;
    user.actions.getUser = () => Promise.resolve();
  });

  afterAll(() => {
    user.actions.getUser = originGetUser;
  });

  describe('snapshots', () => {
    it('should match snapshot', () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('props', () => {
    it('should transfer isAuthLoading props', () => {
      const userCopy = clone(user);
      const isAuthLoading = true;

      userCopy.state.isAuthLoading = isAuthLoading;
      const { wrapper } = createWrapper({
        modules: { user: userCopy },
        stubs: { AuthForm },
      });

      expect(wrapper.find(AuthForm).props().loading).toBe(isAuthLoading);
    });

    it('should transfer error props', () => {
      const userCopy = clone(user);
      const error = 'Some server error';

      userCopy.state.error = error;
      const { wrapper } = createWrapper({
        modules: { user: userCopy },
        stubs: { AuthForm },
      });

      expect(wrapper.find(AuthForm).props().error).toBe(error);
    });
  });

  describe('events', () => {
    it('should on submit event call auth action', () => {
      const userCopy = clone(user);
      const payload = { login: 'user@gmail.com', password: 'sam001', type: 'sign-in' };
      const spy = jest.fn();

      userCopy.actions.auth = spy;
      const { wrapper } = createWrapper({
        modules: { user: userCopy },
        stubs: { AuthForm },
      });

      wrapper.find(AuthForm).vm.$emit('submit', payload);
      expect(spy.mock.calls[0][1]).toBe(payload);
    });

    it('should on toggle-type event call clearError action', () => {
      const userCopy = clone(user);
      const payload = 'sign-up';
      const spy = jest.fn();

      userCopy.actions.clearError = spy;
      const { wrapper } = createWrapper({
        modules: { user: userCopy },
        stubs: { AuthForm },
      });

      wrapper.find(AuthForm).vm.$emit('toggle-type', payload);
      expect(spy.mock.calls[0][1]).toBe(payload);
    });
  });

  describe('hooks', () => {
    it('should call getUser action after mount', () => {
      const userCopy = clone(user);
      const spy = jest.fn();

      userCopy.actions.getUser = spy;
      createWrapper({ modules: { user: userCopy }, isWithoutStubs: true });

      expect(spy).toHaveBeenCalled();
    });
  });
});
