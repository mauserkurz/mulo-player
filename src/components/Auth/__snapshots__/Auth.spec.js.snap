// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Auth snapshots should match snapshot 1`] = `
<main class="v-main auth" style="padding: 0px 0px 0px 0px;">
  <div class="v-main__wrap">
    <div class="container fill-height container--fluid">
      <div class="row align-center justify-center">
        <div class="col-sm-8 col-md-4 col-12">
          <div class="auth-form elevation-12 v-card v-sheet theme--light">
            <form novalidate="novalidate" class="v-form">
              <header class="v-sheet theme--dark v-toolbar v-toolbar--flat primary" style="height: 64px;">
                <div class="v-toolbar__content" style="height: 64px;">
                  <div class="v-toolbar__title">Login form</div>
                  <div class="spacer"></div> <button type="button" class="auth-form__sign-in-tab v-btn v-btn--flat v-btn--text theme--dark v-size--small"><span class="v-btn__content">
        Sign In
      </span></button> <button type="button" class="auth-form__sign-up-tab v-btn v-btn--flat v-btn--text theme--dark v-size--small"><span class="v-btn__content">
        Sign Up
      </span></button>
                </div>
              </header>
              <div class="v-card__text">
                <div class="v-input theme--light v-text-field">
                  <div class="v-input__prepend-outer">
                    <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">person</i></div>
                  </div>
                  <div class="v-input__control">
                    <div class="v-input__slot">
                      <div class="v-text-field__slot"><label for="input-12" class="v-label theme--light" style="left: 0px; position: absolute;">Login</label><input name="login" id="input-12" type="email"></div>
                    </div>
                    <div class="v-text-field__details">
                      <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
                    </div>
                  </div>
                </div>
                <div class="v-input theme--light v-text-field">
                  <div class="v-input__prepend-outer">
                    <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">lock</i></div>
                  </div>
                  <div class="v-input__control">
                    <div class="v-input__slot">
                      <div class="v-text-field__slot"><label for="input-16" class="v-label theme--light" style="left: 0px; position: absolute;">Password</label><input name="password" id="input-16" type="password"></div>
                    </div>
                    <div class="v-text-field__details">
                      <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
                    </div>
                  </div>
                </div>
              </div>
              <div class="v-card__text" style="display: none;">
                <div role="alert" class="v-alert v-sheet theme--dark v-alert--dense error">
                  <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--dark"></i>
                    <div class="v-alert__content">

                    </div>
                  </div>
                </div>
              </div>
              <div class="v-card__actions">
                <div class="spacer"></div> <button type="submit" class="auth-form__submit-button v-btn v-btn--contained theme--light v-size--default primary"><span class="v-btn__content">
        Login
      </span></button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</main>
`;
